### PR TITLE
Add order model

### PR DESCRIPTION
**Pull request by Yam0r**

### CODE REVIEW

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
resolve branch conflicts first

File: src/main/java/mate/academy/springintro/config/OrderMapper.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import mate.academy.springintro.model.Order;
+import mate.academy.springintro.model.OrderItem;
+import mate.academy.springintro.model.OrderStatus;
+import mate.academy.springintro.model.User;
+import mate.academy.springintro.service.UserService;
+import org.springframework.stereotype.Component;
+
+@AllArgsConstructor
+@Component
+public class OrderMapper {
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/config/OrderMapper.java]: Move the logic of creating an order to the service class and add method `toDto` here. Implement this mapper like the others. 
By the way, it's a bad practice to load a data from the DB in the mapper.

File: src/main/java/mate/academy/springintro/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    @PostMapping
+    @Operation(summary = "Place a new order",
+            description = "Place a new order with items and shipping address")
+    public ResponseEntity<Order> placeOrder(@RequestBody @Valid OrderRequestDto orderRequestDto,
+                                            Authentication authentication) {
+        Long userId = getUserIdFromAuthentication(authentication);
+        Set<OrderItem> orderItems = getOrderItemsFromRequest(orderRequestDto);
+        Order order = orderService.placeOrder(userId,
+                orderRequestDto.getShippingAddress(), orderItems);
+        return ResponseEntity.status(HttpStatus.CREATED).body(order);
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/controller/OrderController.java]: Something strange happens here)
```suggestion
        User user = (User) authentication.getPrincipal();
        OrderResponseDto orderResponseDto = orderService.placeOrder(user, orderRequestDto);
        return ResponseEntity.status(HttpStatus.CREATED).body(orderResponseDto);
```

File: src/main/java/mate/academy/springintro/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    @PreAuthorize("hasAuthority('USER') or hasAuthority('ADMIN')")
+    @GetMapping
+    @Operation(summary = "Get order history",
+            description = "Retrieve a list of all orders made by the user")
+    public ResponseEntity<List<Order>> getOrderHistory(Authentication authentication,
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/controller/OrderController.java]: Same here. We need to have all the logic at the service layer + return a DTO object to the client, not an entity

File: src/main/java/mate/academy/springintro/controller/OrderController.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+@AllArgsConstructor
+public class OrderController {
+
+    private final OrderService orderService;
+
+    @PreAuthorize("hasAuthority('USER') or hasAuthority('ADMIN')")
+    @PostMapping
+    @Operation(summary = "Place a new order",
+            description = "Place a new order with items and shipping address")
+    public ResponseEntity<Order> placeOrder(@RequestBody @Valid OrderRequestDto orderRequestDto,
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/controller/OrderController.java]:
```suggestion
    public ResponseEntity<OrderResponseDto> placeOrder(@RequestBody @Valid OrderRequestDto orderRequestDto,
```

File: src/main/java/mate/academy/springintro/controller/OrderController.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        return ResponseEntity.ok(order);
+    }
+
+    private Long getUserIdFromAuthentication(Authentication authentication) {
+        return Long.valueOf(authentication.getName());
+    }
+
+    private Set<OrderItem> getOrderItemsFromRequest(OrderRequestDto orderRequestDto) {
+        return orderRequestDto.getOrderItems();
+    }
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/controller/OrderController.java]:
```suggestion
```

File: src/main/java/mate/academy/springintro/controller/OrderController.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        Long userId = getUserIdFromAuthentication(authentication);
+        List<Order> orders = orderService.getUserOrders(userId, pageable);
+        return ResponseEntity.ok(orders);
+    }
+
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @PatchMapping("/{id}")
+    @Operation(summary = "Update order status",
+            description = "Update the status of an existing order by its ID")
+    public ResponseEntity<Order> updateOrderStatus(@PathVariable Long id,
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/controller/OrderController.java]: Same

File: src/main/java/mate/academy/springintro/model/Order.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+@Entity
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted = FALSE")
+@Table(name = "orders")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/model/Order.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/academy/springintro/model/OrderItem.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+@Entity
+@SQLDelete(sql = "UPDATE order_items SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted = FALSE")
+@Table(name = "order_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/model/OrderItem.java]: Same here

File: src/main/java/mate/academy/springintro/model/OrderItem.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+
+    @ManyToOne
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/model/OrderItem.java]: Same here

File: src/main/java/mate/academy/springintro/model/OrderItem.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Getter
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE order_items SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted = FALSE")
+@Table(name = "order_items")
+public class OrderItem {
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/model/OrderItem.java:20]: In `Order` you have `Set<OrderItem>`, but you didn't override `equals()` and `hashCode()` methods. It potentially can cause bugs.
Here are 2 articles on how to do this (we need to use the approach like in the second one):
https://vladmihalcea.com/hibernate-facts-equals-and-hashcode/
https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier

File: src/main/java/mate/academy/springintro/service/UserService.java [hunk 11]
```diff
@@ -3,7 +3,10 @@
 import mate.academy.springintro.dto.user.UserRegistrationRequestDto;
 import mate.academy.springintro.dto.user.UserResponseDto;
 import mate.academy.springintro.exception.RegistrationException;
+import mate.academy.springintro.model.User;
 
 public interface UserService {
     UserResponseDto register(UserRegistrationRequestDto requestDto) throws RegistrationException;
+
+    User findById(Long id);
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/service/UserService.java]: You can pass a `User` from the controller
```suggestion
```

File: src/main/java/mate/academy/springintro/service/OrderService.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package mate.academy.springintro.service;
+
+import java.util.List;
+import java.util.Set;
+import mate.academy.springintro.model.Order;
+import mate.academy.springintro.model.OrderItem;
+import mate.academy.springintro.model.OrderStatus;
+import org.springframework.data.domain.Pageable;
+
+public interface OrderService {
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/service/OrderService.java:11]: You have to return DTOs, not entites

File: src/main/java/mate/academy/springintro/service/impl/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+@Service
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final OrderMapper orderMapper;
+
+    @Override
+    public Order placeOrder(Long userId, String shippingAddress, Set<OrderItem> orderItems) {
+        OrderRequestDto orderRequestDto = new OrderRequestDto(shippingAddress, orderItems);
+        Order order = orderMapper.toOrder(userId, orderRequestDto, orderItems);
+        return orderRepository.save(order);
```
Comment (JJJazl) [inline src/main/java/mate/academy/springintro/service/impl/OrderServiceImpl.java]: This method should be completely rewritten, let's discuss it in the chat or Q&A session if you will have questions

**Review by JJJazl — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/springintro/controller/OrderController.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    @PatchMapping("/{id}")
+    @Operation(summary = "Update order status",
+            description = "Update the status of an existing order by its ID")
+    public OrderResponseDto updateOrderStatus(@PathVariable Long id,
+                                              @RequestBody @Valid UpdateOrderStatusRequest
+                                                      updateRequest) {
+        return orderService.updateOrderStatus(id, updateRequest);
+    }
+
+    private Long getUserIdFromAuthentication(Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/controller/OrderController.java]: it looks like unused method

File: src/main/java/mate/academy/springintro/controller/OrderController.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+        return Long.valueOf(authentication.getName());
+    }
+
+    private Set<OrderItem> getOrderItemsFromRequest(OrderRequestDto orderRequestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/controller/OrderController.java]: remove unused method

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by sarakhmen — APPROVED** [review]
lgtm!

**Pull request merged**